Moving/Updating the WasmPlugin tutorial and Adding ImagePullPolicy description (#11379)
istio/istio.io: content/en/docs/ops/configuration/extensibility/wasm-pull-policy/index.md

---
title: Pull Policy for WebAssembly Modules
description: Describes how to determine whether newly pulling Wasm modules or using the cached one.
weight: 10
aliases:
  - /help/ops/extensibility/wasm-pull-policy
  - /docs/ops/extensibility/wasm-pull-policy
keywords: [extensibility,Wasm,WebAssembly]
owner: istio/wg-policies-and-telemetry-maintainers
test: n/a
status: Alpha
---

[WasmPlugin API](/docs/reference/config/proxy_extensions/wasm-plugin) provides the way of [distributing Wasm module](/docs/tasks/extensibility/wasm-module-distribution) to proxies.
Since each proxy should pull Wasm modules from remote registry or HTTP server, understanding of the pull policy would be important in terms of usability as well as performance.

## Image pull policy and exceptions

Inspired from `ImagePullPolicy` of Kubernetes, [WasmPlugin](/docs/reference/config/proxy_extensions/wasm-plugin/#WasmPlugin) also has the notion of `IfNotPresent` and `Always`, which means "use the cached module" and "always pull the module regardless of the cache", respectively.

> By using `ImagePullPolicy` field, the user can configure the policy explicitly except following cases:

Reviewer: ```suggestion
Users explicitly configure the behavior for Wasm module retrieval with the `ImagePullPolicy` field. However, user-provided behavior can be overridden by Istio in the following scenarios:
```

1. If the user sets `sha256` in [WasmPlugin](/docs/reference/config/proxy_extensions/wasm-plugin/#WasmPlugin), regardless of `ImagePullPolicy`, `IfNotPresent` policy is used.

> 1. If `url` field is pointing an OCI image and it has a digest suffix (e.g., `gcr.io/foo/bar@sha256:0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef`), `IfNotPresent` policy is used.

Reviewer: ```suggestion
1. If the `url` field points to an OCI image and it has a digest suffix (e.g., `gcr.io/foo/bar@sha256:0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef`), `IfNotPresent` policy is used.
```

> 1. Similar with Kubernetes, when `ImagePullPolicy` is not specified, `IfNotPresent` is assumed basically. But, in case that `url` field specifies an OCI image having `latest` tag, `Always` policy is used.

Reviewer: ```suggestion

When `ImagePullPolicy` is not specified for a resource, Istio defaults to `IfNotPresent` behavior. However, if the provided `url` field specifies an OCI image that has a tag value of `latest`, Istio will use `Always` behavior.
```

## Life of cached modules

Each proxy, which is a sidecar proxy or gateway, caches Wasm modules. Therefore, lifetime of the cached Wasm module is bounded by the lifetime of the corresponding pod.
In addition, there is expiration mechanism for keeping small footprint of memory: if a cached Wasm module is not used for the certain amount of the time, the Wasm module is purged.

The expiration can be configured via the environment variables `WASM_MODULE_EXPIRY` and `WASM_PURGE_INTERVAL` of [pilot-proxy](/docs/reference/commands/pilot-agent/#envvars), which are the duration of expiration and the interval for checking the expiration.

## Meaning of Always

> In Kubernetes, meaning of `Always` is "every time for spawning new pod".

Reviewer: ```suggestion
In Kubernetes, `ImagePullPolicy: Always` means that an image is pulled directly from its source each time a pod is created.
```

> So, each time of spawning new pod, Kubernetes newly pulls the image.

Reviewer: ```suggestion
Every time of a new pod is spanned, Kubernetes pulls the image anew.
```

> In case of `WasmPlugin`, the meaning of `Always` is "every time for creating/changing the corresponding `WasmPlugin` resource".

Reviewer: ```suggestion
In case of `WasmPlugin`, `ImagePullPolicy: Always` means that Istio will pull an image directly from its source each time the corresponding `WasmPlugin` Kubernetes resource is created or changed.
```

> Please note that a change not only in `spec` but also `metadata` triggers pulling of Wasm module when `Always` policy is used.

Reviewer: ```suggestion
Please note that a change not only in `spec` but also `metadata` triggers pulling of Wasm module when `Always` policy is used. This can mean that an image is pulled from source several times over the lifetime of a pod and over the lifetime of an individual proxy.
```

---
title: Pull Policy for WebAssembly Modules

> description: Describes how to determine whether newly pulling Wasm modules or using the cached one.

Reviewer 2: ```suggestion
description: Describes how Istio determines whether to pull Wasm modules or use cached versions.
```

weight: 10

> aliases:
>   - /help/ops/extensibility/wasm-pull-policy
>   - /docs/ops/extensibility/wasm-pull-policy

Reviewer 2: ```suggestion
```
never had another location so can remove these

keywords: [extensibility,Wasm,WebAssembly]
owner: istio/wg-policies-and-telemetry-maintainers
test: n/a
status: Alpha
---

> [WasmPlugin API](/docs/reference/config/proxy_extensions/wasm-plugin) provides the way of [distributing Wasm module](/docs/tasks/extensibility/wasm-module-distribution) to proxies.
> Since each proxy should pull Wasm modules from remote registry or HTTP server, understanding of the pull policy would be important in terms of usability as well as performance.

Reviewer 2: ```suggestion
The [WasmPlugin API](/docs/reference/config/proxy_extensions/wasm-plugin) provides a method for [distributing Wasm modules](/docs/tasks/extensibility/wasm-module-distribution) to proxies.
Since each proxy will pull Wasm modules from a remote registry or an HTTP server, understanding how Istio chooses to pull modules is important in terms of usability as well as performance.
```

## Image pull policy and exceptions

> Inspired from `ImagePullPolicy` of Kubernetes, [WasmPlugin](/docs/reference/config/proxy_extensions/wasm-plugin/#WasmPlugin) also has the notion of `IfNotPresent` and `Always`, which means "use the cached module" and "always pull the module regardless of the cache", respectively.

Reviewer 3: I'd opt for `Inspired by` or `Analogous to`

Users explicitly configure the behavior for Wasm module retrieval with the `ImagePullPolicy` field. However, user-provided behavior can be overridden by Istio in the following scenarios:

1. If the user sets `sha256` in [WasmPlugin](/docs/reference/config/proxy_extensions/wasm-plugin/#WasmPlugin), regardless of `ImagePullPolicy`, `IfNotPresent` policy is used.
1. If the `url` field points to an OCI image and it has a digest suffix (e.g., `gcr.io/foo/bar@sha256:0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef`), `IfNotPresent` policy is used.

When `ImagePullPolicy` is not specified for a resource, Istio defaults to `IfNotPresent` behavior. However, if the provided `url` field specifies an OCI image that has a tag value of `latest`, Istio will use `Always` behavior.

> ## Life of cached modules
> 
> Each proxy, which is a sidecar proxy or gateway, caches Wasm modules. Therefore, lifetime of the cached Wasm module is bounded by the lifetime of the corresponding pod.
> In addition, there is expiration mechanism for keeping small footprint of memory: if a cached Wasm module is not used for the certain amount of the time, the Wasm module is purged.
> 
> The expiration can be configured via the environment variables `WASM_MODULE_EXPIRY` and `WASM_PURGE_INTERVAL` of [pilot-proxy](/docs/reference/commands/pilot-agent/#envvars), which are the duration of expiration and the interval for checking the expiration.

Reviewer 2: ```suggestion
## Lifecycle of cached modules

Each proxy, whether a sidecar proxy or a gateway, caches Wasm modules. The lifetime of the cached Wasm module is therefore bounded by the lifetime of the corresponding pod.
In addition, there is an expiration mechanism for keeping the proxy memory footprint to a minimum: if a cached Wasm module is not used for a certain amount of the time, the module is purged.

This expiration can be configured via the environment variables `WASM_MODULE_EXPIRY` and `WASM_PURGE_INTERVAL` of [pilot-proxy](/docs/reference/commands/pilot-agent/#envvars), which are the duration of expiration and the interval for checking the expiration respectively.
```

## Meaning of Always

In Kubernetes, `ImagePullPolicy: Always` means that an image is pulled directly from its source each time a pod is created.

> Every time of a new pod is spanned, Kubernetes pulls the image anew.

Reviewer 3: `spawned`

> Every time of a new pod is spanned, Kubernetes pulls the image anew.

Reviewer 3: also maybe something like

```suggestion
Every time of a new pod is spanned, Kubernetes pulls the image anew. As istio does not use your cluster's container runtime to pull the image, subsequent pulls on the same node will not be faster than the first. Istio will always pull the full image, no matter if it has pulled an image with the same SHA previously. Keep this in mind when using `ImagePullPolicy: Always`. 
```

Author: Oh... I understood.

Thing is, in current implementation, we check the manifest first, and then determine whether to pull or just to use the cached Wasm binary. So, the difference with K8S is only that they are cached in "proxy instance" not "node".

Since this is explained in the previous section, IMHO, it seems to be not necessary to add this part.

WDYT?

Author: At this time, let's go with the original sentence.

@dgn Anytime, please let me know if you have any concern here. Thanks!

> Every time of a new pod is spanned, Kubernetes pulls the image anew.

Author: ```suggestion
Every time of a new pod is spawned, Kubernetes pulls the image anew.
```

> Every time of a new pod is spanned, Kubernetes pulls the image anew.

Author: ```suggestion
Every time of a new pod is spawned, Kubernetes pulls the image anew.
```

> ## Meaning of Always
> 
> In Kubernetes, `ImagePullPolicy: Always` means that an image is pulled directly from its source each time a pod is created.
> Every time of a new pod is spanned, Kubernetes pulls the image anew.
> 
> In case of `WasmPlugin`, `ImagePullPolicy: Always` means that Istio will pull an image directly from its source each time the corresponding `WasmPlugin` Kubernetes resource is created or changed.
> Please note that a change not only in `spec` but also `metadata` triggers pulling of Wasm module when `Always` policy is used. This can mean that an image is pulled from source several times over the lifetime of a pod and over the lifetime of an individual proxy.

Reviewer 2: ```suggestion
## The meaning of "Always"

In Kubernetes, `ImagePullPolicy: Always` means that an image is pulled directly from its source each time a pod is created.
Every time a new pod is started, Kubernetes pulls the image anew.

For a `WasmPlugin`, `ImagePullPolicy: Always` means that Istio will pull an image directly from its source each time the corresponding `WasmPlugin` Kubernetes resource is created or changed.
Please note that a change not only in `spec` but also `metadata` triggers the pulling of a Wasm module when the `Always` policy is used. This can mean that an image is pulled from source several times over the lifetime of a pod, and over the lifetime of an individual proxy.
```
